pkg/header/check.go

```diff
@@ -28,92 +30,144 @@ import (
 	lcs "github.com/apache/skywalking-eyes/pkg/license"
 
 	"github.com/bmatcuk/doublestar/v2"
+	"github.com/go-git/go-billy/v5/osfs"
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing/format/gitignore"
+	"github.com/go-git/go-git/v5/plumbing/object"
 )
 
 // Check checks the license headers of the specified paths/globs.
 func Check(config *ConfigHeader, result *Result) error {
-	for _, pattern := range config.Paths {
-		if err := checkPattern(pattern, result, config); err != nil {
+	fileList, err := listFiles(config)
+	if err != nil {
+		return err
+	}
+
+	for _, file := range fileList {
+		if err := CheckFile(file, config, result); err != nil {
 			return err
 		}
 	}
 
 	return nil
 }
 
-var seen = make(map[string]bool)
+func listFiles(config *ConfigHeader) ([]string, error) {
+	var fileList []string
 
-func checkPattern(pattern string, result *Result, config *ConfigHeader) error {
-	paths, err := glob(pattern)
+	repo, err := git.PlainOpen("./")
 
-	if err != nil {
-		return err
-	}
+	if err != nil { // we're not in a Git workspace, fallback to glob paths
+		var localFileList []string
+		for _, pattern := range config.Paths {
+			if pattern == "." {
+				pattern = "./"
+			}
+			files, err := doublestar.Glob(pattern)
+			if err != nil {
+				return nil, err
+			}
+			localFileList = append(localFileList, files...)
+		}
+
+		seen := make(map[string]bool)
+		for _, file := range localFileList {
+			files, err := walkFile(file, seen)
+			if err != nil {
+				return nil, err
+			}
+			fileList = append(fileList, files...)
+		}
+	} else {
+		var candidates []string
 
-	for _, path := range paths {
-		if yes, err := config.ShouldIgnore(path); yes || err != nil {
-			result.Ignore(path)
-			continue
+		t, _ := repo.Worktree()
+		if t.Excludes == nil {
+			t.Excludes = make([]gitignore.Pattern, 0)
 		}
-		if err := checkPath(path, result, config); err != nil {
-			return err
+		if ignorePattens, err := gitignore.LoadGlobalPatterns(osfs.New("")); err == nil {
+			t.Excludes = append(t.Excludes, ignorePattens...)
+		}
+		if ignorePattens, err := gitignore.LoadSystemPatterns(osfs.New("")); err == nil {
+			t.Excludes = append(t.Excludes, ignorePattens...)
+		}
+		s, _ := t.Status()
+		for file := range s {
+			candidates = append(candidates, file)
 		}
-		seen[path] = true
-	}
 
-	return nil
-}
+		head, _ := repo.Head()
+		commit, _ := repo.CommitObject(head.Hash())
+		tree, err := commit.Tree()
+		if err != nil {
+			return nil, err
+		}
+		if err := tree.Files().ForEach(func(file *object.File) error {
+			if file == nil {
+				return errors.New("file pointer is nil")
+			}
+			if _, err := os.Stat(file.Name); err == nil {
+				candidates = append(candidates, file.Name)
+			}
+			return nil
```

Comment:
I prefer to keep the previous logic so when `os.Stat` reports an error we won't leave out a file, (we don't care the error, tho). 

```suggestion
			if _, err := os.Stat(file.Name); err != nil {
				return err
			}
			candidates = append(candidates, file.Name)
			return nil
```

Reply:
It's possible that a file is removed in staging but not committed. Thus, files listed by `git ls-files` or `git status` can be deleted. It's not an error case.

It's possible we skip return error if `os.IsNotExist(err) == true`.

WDYT?

Reply:
> It's possible that a file is removed in staging but not committed. Thus, files listed by `git ls-files` or `git status` can be deleted. It's not an error case.
> 
> It's possible we skip return error if `os.IsNotExist(err) == true`.
> 
> WDYT?

OK I agree, let's add the if condition.

Reply:
Updated at https://github.com/apache/skywalking-eyes/pull/133/commits/8565a1f85bc34debafd764a39cc1c5bb287ba0b7.